lib/vm/src/instance.rs

```diff
@@ -761,99 +680,98 @@ impl Instance {
-/// TODO: this needs extra review
-unsafe impl Send for InstanceHandle {}
+impl InstanceAllocator {
+    #[inline(always)]
+    const fn new(instance: Instance, ptr_to_dealloc: *const Arc<Self>, ptr_layout: Layout) -> Self {
+        Self {
+            instance: mem::ManuallyDrop::new(instance),
+            ptr_to_dealloc,
+            ptr_layout,
+        }
+    }
 
-impl InstanceHandle {
-    /// Allocates an instance for use with `InstanceHandle::new`.
-    ///
-    /// Returns the instance pointer and the [`VMOffsets`] that describe the
-    /// memory buffer pointed to by the instance pointer.
-    pub fn allocate_instance(module: &ModuleInfo) -> (NonNull<u8>, VMOffsets) {
-        let offsets = VMOffsets::new(mem::size_of::<*const u8>() as u8, module);
+    fn instance_layout(offsets: &VMOffsets) -> Layout {
+        let size = mem::size_of::<Arc<mem::ManuallyDrop<Self>>>()
+            .checked_add(usize::try_from(offsets.size_of_vmctx()).unwrap())
+            .unwrap();
+        let align = mem::align_of::<Arc<mem::ManuallyDrop<Self>>>();
+
+        Layout::from_size_align(size, align).unwrap()
+    }
 
-        let layout = Instance::alloc_layout(&offsets);
+    fn allocate_instance(offsets: &VMOffsets) -> NonNull<u8> {
+        let layout = Self::instance_layout(offsets);
 
         #[allow(clippy::cast_ptr_alignment)]
-        let instance_ptr = unsafe { alloc::alloc(layout) as *mut Instance };
-        let ptr = if let Some(ptr) = NonNull::new(instance_ptr) {
+        let arc_instance_ptr = unsafe { alloc::alloc(layout) as *mut Arc<InstanceAllocator> };
+
+        let ptr = if let Some(ptr) = NonNull::new(arc_instance_ptr) {
             ptr.cast()
         } else {
             alloc::handle_alloc_error(layout);
         };
 
-        (ptr, offsets)
+        ptr
     }
 
-    /// Get the locations of where the local `VMMemoryDefinition`s should be stored.
-    ///
-    /// This function lets us create `Memory` objects on the host with backing
-    /// memory in the VM.
-    ///
-    /// # Safety
-    /// - `instance_ptr` must point to enough memory that all of the offsets in
-    ///   `offsets` point to valid locations in memory.
-    pub unsafe fn memory_definition_locations(
-        instance_ptr: NonNull<u8>,
-        offsets: &VMOffsets,
-    ) -> Vec<NonNull<VMMemoryDefinition>> {
-        let num_memories = offsets.num_local_memories;
-        let num_memories = usize::try_from(num_memories).unwrap();
-        let mut out = Vec::with_capacity(num_memories);
-        // TODO: better encapsulate this logic, this shouldn't be duplicated
-        let base_ptr = instance_ptr.as_ptr().add(std::mem::size_of::<Instance>());
-        for i in 0..num_memories {
-            let mem_offset = offsets.vmctx_vmmemory_definition(LocalMemoryIndex::new(i));
-            let mem_offset = usize::try_from(mem_offset).unwrap();
+    unsafe fn deallocate_instance(&mut self) {
+        mem::ManuallyDrop::drop(&mut self.instance);
```

Comment:
TODO: review this very closely, this looks like it's dropping the wrong thing (the pointer, not the thing it's pointing to)

Reply:
This has radically changed. The commented code has been removed.